### PR TITLE
shader_recompiler: Implement most integer image atomics, workgroup barriers and shared memory load/store

### PR DESCRIPTION
**Pull request by raphaelthegreat**
Couldn't resist not testing this since it showed intro screens lol. These allow Red Dead Redemption to not crash during shader compilation. The game also uses inline constant buffers but those should be for a different PR.

For shared memory emulation we use VK_KHR_workgroup_memory_explicit_layout to allow aliasing of the shared region with different access types, which makes reads/writes with granularity less than a dword easy.

Also fixed intro text by allowing render target refresh in texture cache. This is not ideal as it will upload from host memory each time but it works. Will need a proper guest buffer manager to make it fast

### CODE REVIEW

**Comment by yaya54840** [comment]
Hi View raphaelthegreat Thanks for the work you do.
I am falling back on your emulator because fpps4 and psOff are stopped.
As soon as I launch a game with the latest Windows Qt it returns to desktop directly.
I put back an old shadps4.exe for proper operation.
I have win 11 pro with AMD 9600 and 7800XT graphics card.

**Comment by georgemoralis** [comment]
this is not a place for bug report .Open an issue instead

File: src/core/libraries/kernel/memory_management.cpp [hunk 1]
```diff
@@ -110,10 +110,13 @@ int PS4_SYSV_ABI sceKernelMapNamedDirectMemory(void** addr, u64 len, int prot, i
         }
     }
 
-    const VAddr in_addr = reinterpret_cast<VAddr>(*addr);
+    VAddr in_addr = reinterpret_cast<VAddr>(*addr);
     const auto mem_prot = static_cast<Core::MemoryProt>(prot);
     const auto map_flags = static_cast<Core::MemoryMapFlags>(flags);
     auto* memory = Core::Memory::Instance();
+    if (False(map_flags & Core::MemoryMapFlags::Fixed) && in_addr == 0) {
```
Comment (georgemoralis) [inline src/core/libraries/kernel/memory_management.cpp]: leftover?

File: src/core/libraries/kernel/memory_management.cpp [hunk 2]
```diff
@@ -143,10 +146,13 @@ s32 PS4_SYSV_ABI sceKernelMapNamedFlexibleMemory(void** addr_in_out, std::size_t
         return ORBIS_KERNEL_ERROR_EFAULT;
     }
 
-    const VAddr in_addr = reinterpret_cast<VAddr>(*addr_in_out);
+    VAddr in_addr = reinterpret_cast<VAddr>(*addr_in_out);
     const auto mem_prot = static_cast<Core::MemoryProt>(prot);
     const auto map_flags = static_cast<Core::MemoryMapFlags>(flags);
     auto* memory = Core::Memory::Instance();
+    if (False(map_flags & Core::MemoryMapFlags::Fixed) && in_addr == 0) {
```
Comment (georgemoralis) [inline src/core/libraries/kernel/memory_management.cpp]: leftover?

File: src/core/libraries/kernel/file_system.cpp [hunk 3]
```diff
@@ -90,7 +90,7 @@ int PS4_SYSV_ABI sceKernelOpen(const char* path, int flags, u16 mode) {
         }
         if (!file->f.IsOpen()) {
             h->DeleteHandle(handle);
-            return SCE_KERNEL_ERROR_EACCES;
+            return SCE_KERNEL_ERROR_ENOENT;
```
Comment (georgemoralis) [inline src/core/libraries/kernel/file_system.cpp]: proper fix in another pr

File: src/core/libraries/network/net.cpp [hunk 4]
```diff
@@ -59,6 +60,7 @@ int PS4_SYSV_ABI sce_net_in6addr_nodelocal_allnodes() {
 }
 
 OrbisNetId PS4_SYSV_ABI sceNetAccept(OrbisNetId s, OrbisNetSockaddr* addr, u32* paddrlen) {
+    std::this_thread::sleep_for(std::chrono::seconds(60));
```
Comment (georgemoralis) [inline src/core/libraries/network/net.cpp]: hackish

File: src/emulator.cpp [hunk 5]
```diff
@@ -137,8 +137,9 @@ void Emulator::Run(const std::filesystem::path& file) {
 }
 
 void Emulator::LoadSystemModules(const std::filesystem::path& file) {
-    constexpr std::array<SysModules, 6> ModulesToLoad{
+    constexpr std::array<SysModules, 7> ModulesToLoad{
         {{"libSceNgs2.sprx", nullptr},
+         {"libSceFiber.sprx", nullptr},
```
Comment (georgemoralis) [inline src/emulator.cpp]: we haven't review that if it works yet

File: src/video_core/renderer_vulkan/vk_stream_buffer.cpp [hunk 6]
```diff
@@ -226,7 +226,7 @@ void StreamBuffer::WaitPendingOperations(u64 requested_upper_bound) {
     while (requested_upper_bound > wait_bound && wait_cursor < *invalidation_mark) {
         auto& watch = previous_watches[wait_cursor];
         wait_bound = watch.upper_bound;
-        scheduler.Wait(watch.tick);
+        //scheduler.Wait(watch.tick);
```
Comment (georgemoralis) [inline src/video_core/renderer_vulkan/vk_stream_buffer.cpp]: hmm?

**Pull request merged**